tools/rule-notices.ts

```diff
@@ -0,0 +1,109 @@
+import plugin from '../src';
+
+enum MESSAGE_TYPE {
+  CONFIGS = 1,
+  DEPRECATED = 2,
+  FIXABLE = 3,
+  HAS_SUGGESTIONS = 4,
+}
+
+export const MESSAGES = {
+  [MESSAGE_TYPE.CONFIGS]:
+    '💼 This rule is enabled in the following [configs](https://github.com/jest-community/eslint-plugin-jest#shareable-configurations):',
+  [MESSAGE_TYPE.DEPRECATED]: '❌ This rule is deprecated.',
+  [MESSAGE_TYPE.FIXABLE]:
+    '🔧 This rule is automatically fixable using the `--fix` [option](https://eslint.org/docs/latest/user-guide/command-line-interface#--fix) on the command line.',
+  [MESSAGE_TYPE.HAS_SUGGESTIONS]:
+    '💡 This rule provides [suggestions](https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions) that can be applied manually.',
+};
+
+type RuleName = keyof typeof plugin.rules;
+type Rule = typeof plugin.rules[RuleName];
+
+/**
+ * Get config names that a given rule belongs to.
+ */
+function getConfigsForRule(ruleName: RuleName) {
+  const { configs } = plugin;
+  const configNames: Array<keyof typeof configs> = [];
+  let configName: keyof typeof configs;
+
+  for (configName in configs) {
+    const config = configs[configName];
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment -- we don't have a static type for rule names
+    // @ts-ignore
+    const value = config.rules[`jest/${ruleName}`];
+    const isEnabled = [2, 'error'].includes(value);
+
+    if (isEnabled) {
+      configNames.push(configName);
+    }
+  }
+
+  return configNames.sort();
+}
+
+/**
+ * Convert list of configs to formatted string list.
+ */
+function configNamesToList(configNames: string[]) {
+  return `\`${configNames.join('`, `')}\``;
+}
+
+/**
+ * Determine which notices should and should not be included at the top of a rule doc.
+ */
+export function getNoticesForRule(rule: Rule) {
```

Comment:
I'm hoping that by moving the handling of notices entirely into the tool we can re-implement this to leverage native data structures a bit more rather than need all these elses i.e. we could have a map of `{ message: expected }`:

```
{
  [MESSAGE_TYPE.DEPRECATED]: rule.meta.deprecated,
}
```

which you could iterate over with something like:

```
for (const [notice, expected] of Object.entries(notices)) {
  if (expected) {
    header.push(notice);

    continue;
  }

  // or maybe linesToRemove.push(notice), and then do that in the outer function
  content.replace(notice, '');
}
```

(this is very psudo-y; let me know if that doesn't make sense 😅)

Reply:
I switched over to having `getNoticesForRule()` return an object. I agree this is much nicer.